Alerting: Support hysteresis command expression

yuri-tceretian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**
This PR adds support for a new type of command introduced in https://github.com/grafana/grafana/pull/70998. The hysteresis command requires information about the results of the previous evaluation. This PR updates the state manager and scheduler to facilitate this information for the evaluation engine:
- state manager is updated to keep a fingerprint of the result instance for each State. It differs from CacheID or Labelshash because it is calculated from the labels of the raw result.
- scheduler wraps the state manager and the rule into an intermediate structure AlertingResultsFromRuleState and provides it to the rule evaluator.
- The evaluator analyzes the expression tree (parsed rule's query) and if it contains HysteresisCommand, it populates it with the data taken from the AlertingResultsFromRuleState. 

The structure AlertingResultsFromRuleState implements the logic that determines what metric\dimension\state is considered as "loaded":
- state Pending or Alerting
- state reason is empty. This is to avoid execution of exceptional states such as Error\NoData as Alerting.

**Why do we need this feature?**
To support hysteresis expression in alerting. This will help users reduce flapping on the alert rules when the metric value frequently crosses threshold boundaries.

**Who is this feature for?**
Alerting

**Which issue(s) does this PR fix?**:

Related https://github.com/grafana/grafana/issues/6202

**Special notes for your reviewer:**

- PR can be reviewed by commit. In some commits I provide the explanations for chosen approach.
- The majority of the changes are tests. The core backend changes are in just two commits https://github.com/grafana/grafana/pull/75189/commits/da01d9ab54c87c1753da90025086c59654764452 and https://github.com/grafana/grafana/pull/75189/commits/ef96be830f7e5cb5e0831a4497af271f6b9f8cb2
- the feature is behind the flag `recoveryThreshold`. if it is disabled the expr engine does not create HysteresisCommand struct and code in this PR does not do anything.

<details><summary>UI screen recording</summary>

https://github.com/grafana/grafana/assets/33540275/06042f40-2f56-46a9-a94a-48ba1afbb0d7

**When viewing the rule** 

<img width="1076" alt="Screenshot 2023-11-06 at 09 45 23" src="https://github.com/grafana/grafana/assets/33540275/5a1eaaa5-cd16-469c-905d-68b8b5c19079">


</details> 

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

yuri-tceretian committed:
Update alerting eval to provide the current results that are alertings to hysteresis command

1. The hysteresis command accepts feedback in the JSON config of the command (consensus with data-metrics team). It is optional, though. I chose to mutate the command after the pipeline is built but before it is executed.
   This allows us to save resources on marshalling back and forth.
2. The instance of the evaluator is run only once by consumer. So, the clean up is not really necessary here. However, I decided to not make this assumption becasue it can change in future.
3. Usage of mutex here is not necessary technically because currently the evaluator instance is not called concurrently. However, in future it may change, e.g. in backtesting.

yuri-tceretian committed:
add rule_fingerprint column to alert_instance

yuri-tceretian committed:
add GetCommandsFromPipeline

yuri-tceretian committed:
add integration test

soniaAguilarPeiron committed:
Add hysteresis option to Threshold in UI

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/expressions/Expression.tsx:
```diff
@@ -57,6 +58,16 @@ export const Expression: FC<ExpressionProps> = ({
+  const onQueriesValidationError = useCallback(
+    (errorMsg: string | undefined) => {
+      errorMsg && setError('queries', { type: 'custom', message: errorMsg });
+      !errorMsg && clearErrors('queries');
```
if we setErrors on the form manually, we need to clear them manually

soniaAguilarPeiron replied:
The reason behind this change is that for Hysteresis expression, we are not adding an actual field in the form but we want to validate it as it was a separate field. 
The only way I found to achieve this,  was calling setError from the FromContext, and as a consequence of this manual setError , we need to clear the error manually.

For more information about how `SetError` works see [documentation](https://react-hook-form.com/docs/useform/seterror)

In rules section you will see:

> An error that is not associated with an input field will be persisted until cleared with clearErrors. This behaviour is only applicable for built-in validation at field level.

yuri-tceretian commented on pkg/services/ngalert/eval/context.go:
```diff
@@ -18,3 +27,11 @@ func NewContext(ctx context.Context, user *user.SignedInUser) EvaluationContext
 	}
 }
+
+func NewContextWithPreviousResults(ctx context.Context, user *user.SignedInUser, reader AlertingResultsReader) EvaluationContext {
```
I created a new factory method because the current one is used in GE

yuri-tceretian committed:
update alerting evaluator to accept AlertingResultsReader via context

Two main benefits:
1. No need to change usages in API handlers because they do not provide the feedback.
2. Simpler to update code in GE.

The new factory method is created to make GE code compilable

yuri-tceretian committed:
add ResultFingerprint to state.State

Different from other fingerprints we store in the state because it is calculated from the result labels.

yuri-tceretian committed:
add tests for alerting evaluator

yuri-tceretian committed:
add AlertingResultsFromRuleState that implements the new interface in eval package

+ update scheduler to provide it to the eval via context

yuri-tceretian committed:
update backtesting engine to use the AlertingResultsFromRuleState

yuri-tceretian commented:
/deploy-to-hg

yuri-tceretian committed:
Merge branch 'up/main' into yuri-tceretian/expr-hysteresis-alerting

yuri-tceretian commented:
/deploy-to-hg

armandgrillet commented:
![SS 2023-10-26 at 06 21 11](https://github.com/grafana/grafana/assets/2117580/485cf1a3-318e-435b-b9e1-6f4b7108b017)
The horizontal alignment looks weird due to switch and label cutting the two. There is a typo (it's `below` not `bellow`).

Here is what I would imagine instead:
![SS 2023-10-26 at 06 26 16](https://github.com/grafana/grafana/assets/2117580/7539bec9-99cb-470c-b8f9-55a6d24121b5)

soniaAguilarPeiron commented:
> ![SS 2023-10-26 at 06 21 11](https://user-images.githubusercontent.com/2117580/278210716-485cf1a3-318e-435b-b9e1-6f4b7108b017.png) The horizontal alignment looks weird due to switch and label cutting the two. There is a typo (it's `below` not `bellow`).
> 
> Here is what I would imagine instead: ![SS 2023-10-26 at 06 26 16](https://user-images.githubusercontent.com/2117580/278211484-7539bec9-99cb-470c-b8f9-55a6d24121b5.png)


thanks, @armandgrillet! you're right, the horizontal alignment it's wrong, let's fix it ! 
But I think moving the switch to the last line it's not a good idea as the expression box becomes too wide.

soniaAguilarPeiron committed:
Address PR comment about alignment and typo

soniaAguilarPeiron commented:
After updating borders to be always visible and fix the aparent misaligment when not hovering: 
 
<img width="637" alt="Screenshot 2023-10-26 at 09 09 27" src="https://github.com/grafana/grafana/assets/33540275/00faab53-11e4-473f-87da-b4f89280044c">


<img width="647" alt="Screenshot 2023-10-26 at 09 09 17" src="https://github.com/grafana/grafana/assets/33540275/359afe43-0009-48eb-8fee-ad1092c86ba0">

soniaAguilarPeiron committed:
Add test for getUnloadEvaluatorTypeFromCondition

soniaAguilarPeiron committed:
Update alignment after an internal discussion with UX

soniaAguilarPeiron commented on public/app/features/expressions/components/Threshold.tsx:
```diff
@@ -476,7 +476,7 @@ const getStyles = (theme: GrafanaTheme2) => ({
     font-size: ${theme.typography.bodySmall.fontSize};
     text-transform: uppercase;
     padding: 0 ${theme.spacing(1)};
-    border: 1px solid ${theme.colors.border.medium};
```
after this change :
<img width="643" alt="Screenshot 2023-10-26 at 10 17 27" src="https://github.com/grafana/grafana/assets/33540275/d701db93-3ac3-45f8-a399-18a6a79bb5b2">

soniaAguilarPeiron committed:
Merge remote-tracking branch 'origin/main' into yuri-tceretian/expr-hysteresis-alerting

soniaAguilarPeiron committed:
Refactor isInvalid and add test for it

soniaAguilarPeiron committed:
Remove unnecesary React.memo

soniaAguilarPeiron committed:
Add some clarifying comments

soniaAguilarPeiron committed:
Add tests for updateEvaluatorConditions

soniaAguilarPeiron commented on public/app/features/expressions/types.ts:
```diff
@@ -140,6 +140,10 @@ export interface ClassicCondition {
     params: number[];
     type: EvalFunction;
   };
+  unloadEvaluator?: {
```
Hysteresis expression is an extension of Threshold expression, with two evaluators (we add the unloadEvaluator)

soniaAguilarPeiron committed:
Address some nits in pr review commments

yuri-tceretian committed:
refactor method GetCommandType to func GetExpressionCommandType

yuri-tceretian committed:
add function to create fingerprint frames

gillesdemey commented on public/app/features/expressions/components/Threshold.tsx:
```diff
@@ -105,48 +206,290 @@ export const Threshold = ({ labelWidth, onChange, refIds, query }: Props) => {
+  }>,
+  hysteresisChecked: boolean
+): ClassicCondition[] {
+  const defaultUnloaEvaluator = {
```
dito typo; `defaultUnloadEvaluator` :)

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/rule-editor/alert-rule-form/AlertRuleForm.tsx:
```diff
@@ -109,7 +109,6 @@ export const AlertRuleForm = ({ existing, prefill }: Props) => {
       notifyApp.error(conditionErrorMsg);
       return;
     }
-
```
yes, linter did. 💯 keeping the diff clean 👍

soniaAguilarPeiron commented on public/app/features/expressions/components/Threshold.tsx:
```diff
@@ -105,48 +205,290 @@ export const Threshold = ({ labelWidth, onChange, refIds, query }: Props) => {
-      evaluator: {
-        ...conditions[0].evaluator,
-        ...update,
+  const hsyteresIsChecked = Boolean(conditions[0].unloadEvaluator);
```
good catch, yes!

soniaAguilarPeiron commented on public/app/features/expressions/components/Threshold.tsx:
```diff
@@ -105,48 +206,290 @@ export const Threshold = ({ labelWidth, onChange, refIds, query }: Props) => {
+      },
+    ];
+  } else {
+    const antUnloadEvaluator = conditions[0].unloadEvaluator ?? defaultUnloaEvaluator;
```
again 🤣

soniaAguilarPeiron commented on public/app/features/expressions/components/Threshold.tsx:
```diff
@@ -48,24 +100,73 @@ export const Threshold = ({ labelWidth, onChange, refIds, query }: Props) => {
+        <InlineFieldRow>
+          <Stack direction="row" gap={2} alignItems="center">
+            <Switch value={hasHysteresis} onChange={onHysteresisCheckChange} />
+            Custom recovery threshold
```
we had a discussion about this misalignments. And we agreed not to use the same row as sometimes this space is not available.

gillesdemey replied:
No, I meant the `InlineSwitch` component, not talking about the switch being on the same line :)

soniaAguilarPeiron replied:
ahh, I was confused with your screenshot having all in the same line 🙂. I finally used the `InlineSwitch` and fixed the misalignments

soniaAguilarPeiron commented on public/app/features/expressions/components/Threshold.tsx:
> I think it would be better to use an inline switch component 
I answered bellow

soniaAguilarPeiron replied:
Final version 👀 
<img width="645" alt="Screenshot 2023-11-03 at 09 38 29" src="https://github.com/grafana/grafana/assets/33540275/cb09ab7a-55e3-460b-9242-9acf34163da8">

soniaAguilarPeiron commented on public/app/features/expressions/components/Threshold.tsx:
```diff
@@ -105,48 +206,290 @@ export const Threshold = ({ labelWidth, onChange, refIds, query }: Props) => {
+    // set error to undefined when type is changed as we default to the new type that is valid
+    onError && onError(undefined); // clear error
+
+    return [
```
> Is there a particular reason why we care if the type has been updated or not? I can't seem to see a reason for it here and we're just appended an unload evaluator when hysteresis is turned on 🤔

yes, if type was changed and hsyteresIsChecked, we need to update the type for the unload evaluator

soniaAguilarPeiron replied:
> Additionally I think this piece of code could also benefit from produce(...) for readability here.

I didn't find a way to use produce in here that makes this code cleaner and more readable 🤔

```
return produce(conditions, draft => {
  draft[0] = {
    ...draft[0],
    evaluator: {
      ...draft[0].evaluator,
      ...update,
    },
    unloadEvaluator: {
      ...defaultUnloadEvaluator,
      type: updateUnloadType,
    },
  };
});
```

yuri-tceretian commented on pkg/services/ngalert/eval/eval.go:
```diff
@@ -67,6 +70,28 @@ func (r *conditionEvaluator) EvaluateRaw(ctx context.Context, now time.Time) (re
+	// patching the hysteresis commands with the current loaded metrics.
+	// This optimization is to save resources on marshalling AlertQuery.
+	// Mutex is to make sure that the evaluation is synchronous.
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	if r.loadedMetricsReader != nil {
+		cmds := expr.GetCommandsFromPipeline[*expr.HysteresisCommand](r.pipeline)
+		if len(cmds) > 0 {
+			loaded := r.loadedMetricsReader.Read()
+			logger.FromContext(ctx).Debug("Populating loaded metrics for hysteresis command", "commands", len(cmds), "loaded", len(loaded))
+			for _, cmd := range cmds {
+				cmd.LoadedDimensions = loaded
+			}
+			// reset the data after evaluation
+			defer func() {
+				for _, cmd := range cmds {
+					cmd.LoadedDimensions = nil
+				}
+			}()
+		}
+	}
```
During the review session the request was to replace this code with one that adds `loadedMetrics` data frame into the query model during conversion from AlertQuery to server-side expressions Request. This will let us reuse the code in the cloud alerting ruler.

yuri-tceretian committed:
revert changes in eval package

yuri-tceretian committed:
move GetModel after datasource type detection

yuri-tceretian committed:
add function to determine whether raw query represents a hysteresis command and a function to patch it with loaded metrics

yuri-tceretian committed:
create methods in AlertQuery to work with hysteresis command

yuri-tceretian committed:
update getExprRequest to patch the hysteresis command

yuri-tceretian committed:
tests

yuri-tceretian committed:
add log

yuri-tceretian committed:
add test to show a bug in frame unmarshalling

yuri-tceretian commented:
PR is blocked by a bug in SDK https://github.com/grafana/grafana-plugin-sdk-go/issues/800

yuri-tceretian committed:
Merge branch 'up/main' into yuri-tceretian/expr-hysteresis-alerting

torkelo committed:
Theme: Update old css class add-data-source-item to use card styling (#77487)

soniaAguilarPeiron commented on public/app/features/expressions/components/Threshold.tsx:
```diff
@@ -1,38 +1,90 @@
+      break;
+    case EvalFunction.IsBelow:
+      if (firstParam < params[0]) {
+        return { errorMsg: `Enter a number more than or equal to ${params[0]}` };
```
That was a suggestion from @dhalachliyski , so let's see what he thinks 👀

soniaAguilarPeiron commented on public/app/features/expressions/components/Threshold.tsx:
```diff
@@ -1,38 +1,90 @@
+      break;
+    case EvalFunction.IsOutsideRange:
+      if (firstParam < params[0]) {
+        return { errorMsgFrom: `Enter a number more than or equal to ${params[0]}` };
```
same, @dhalachliyski 👀

soniaAguilarPeiron commented on public/app/features/expressions/components/Threshold.tsx:
```diff
@@ -1,38 +1,90 @@
+        return { errorMsgFrom: `Enter a number less than or equal to ${params[0]}` };
+      }
+      if (secondParam < params[1]) {
+        return { errorMsgTo: `Enter a number be more than or equal to ${params[1]}` };
```
same ,@dhalachliyski 👀

alexanderzobnin committed:
RolePicker: Use portal to render menu (#77499)

* RolePicker: Use portal for menu

* Remove logging

* Fix submenu styles

* Fix ROLE_PICKER_MAX_MENU_WIDTH calculation

* Fix first menu open glitch

* Fix menu closing on ckick

* Fix menu position

linoman committed:
auth: wire service account proxy (#77215)

* Add interface verification compliance

* rework service account api to a provider

* wire the service accounts api

* rewire the implementation of sa srv for the proxy

---------

Co-authored-by: Misi <mgyongyosi@users.noreply.github.com>

soniaAguilarPeiron committed:
Address pr review comments

soniaAguilarPeiron commented on public/app/features/expressions/components/Threshold.tsx:
```diff
@@ -105,48 +206,290 @@ export const Threshold = ({ labelWidth, onChange, refIds, query }: Props) => {
+  }
+}
+
+function updateUnloadEvaluatorConditions(
```
good idea!

soniaAguilarPeiron committed:
Fix typo

ashharrison90 committed:
Navigation: Use `LoadingBar` in `CommandPalette` (#77506)

try using LoadingBar in CommandPalette

gillesdemey commented:
Let's make sure we don't forget to update the threshold expression viewer to support hysteresis; https://github.com/grafana/grafana/blob/c3962acf9837bc5d659375422acd27db9f2463a6/public/app/features/alerting/unified/GrafanaRuleQueryViewer.tsx#L371-L396

soniaAguilarPeiron commented:
> r to support hysteresis;



good catch!

soniaAguilarPeiron commented:
> Let's make sure we don't forget to update the threshold expression viewer to support hysteresis;

good catch!

soniaAguilarPeiron committed:
Add hysteresis expression in RuleQueryViewer

soniaAguilarPeiron committed:
Merge remote-tracking branch 'origin/main' into yuri-tceretian/expr-hysteresis-alerting

soniaAguilarPeiron committed:
Refactor Threshold component using a reducer to keep the ExpressionQuery data

soniaAguilarPeiron committed:
Merge remote-tracking branch 'origin/main' into yuri-tceretian/expr-hysteresis-alerting

soniaAguilarPeiron committed:
Fix focus being lost in hsyterisis inputs and remove unused props in RecoveryThresholdRow

soniaAguilarPeiron committed:
Refactor

soniaAguilarPeiron committed:
remove spaces

gillesdemey reviewed (approved):
LGTM just one nit – great work on this one! 💪

soniaAguilarPeiron committed:
Remove eslint-disable-next-line comment and add wrap updateQueryExpression in a useCallback to avoid infinite loop in the useEffect

soniaAguilarPeiron committed:
Revert "Remove eslint-disable-next-line comment and add wrap updateQueryExpression in a useCallback to avoid infinite loop in the useEffect"

This reverts commit e01a158a9a154729ce000116a5d9d51057435403.

soniaAguilarPeiron commented on public/app/features/expressions/components/Threshold.tsx:
```diff
@@ -1,70 +1,85 @@
+  // any change in the queryState will trigger the onChange function.
+  useEffect(() => {
+    queryState && onChange(queryState);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
I was unable to remove this one. It appears that `onChange` is being updated each time it's called. While I managed to address it when used in our alerting component by adding a `useCallback` in the parent, I couldn't find a way to implement useCallback when used in the `ExpressionQueryEditor` because `onChange` comes from the `ExpressionDataSource` code. This is the reason for reverting the last commit.

yuri-tceretian committed:
Merge branch 'up/main' into yuri-tceretian/expr-hysteresis-alerting

# Conflicts:
#	pkg/services/ngalert/backtesting/engine.go
#	pkg/services/ngalert/backtesting/engine_test.go
#	pkg/services/ngalert/eval/context.go
#	pkg/services/ngalert/models/testing.go

yuri-tceretian committed:
make assertion sorted

yuri-tceretian committed:
Merge branch 'up/main' into yuri-tceretian/expr-hysteresis-alerting

yuri-tceretian commented:
/deploy-to-hg

soniaAguilarPeiron committed:
Skip validation when unchecking  hysteresis checkbox

soniaAguilarPeiron commented:
last commit : skip hysteresis values being validated when unchecking hysteresis.

https://github.com/grafana/grafana/assets/33540275/5013e272-825d-4fb5-a013-45a2ba5e9685

soniaAguilarPeiron commented:
/deploy-to-hg

yuri-tceretian committed:
Merge branch 'up/main' into yuri-tceretian/expr-hysteresis-alerting

# Conflicts:
#	pkg/expr/threshold_test.go

soniaAguilarPeiron commented:
/deploy-to-hg

soniaAguilarPeiron committed:
Fix hysteresis fields style in dashboard panel

soniaAguilarPeiron commented:
My last commit is a fix for the style for hysteresis fields in dashboard panel expressions.

<details><summary> Before the change:</summary>
 
<img width="971" alt="Screenshot 2023-12-15 at 09 59 07" src="https://github.com/grafana/grafana/assets/33540275/a203eecc-ca7d-4957-bf21-61a845e5004c">
<img width="971" alt="Screenshot 2023-12-15 at 09 58 56" src="https://github.com/grafana/grafana/assets/33540275/e6efde80-3069-48bc-999a-ed5ce7ac8bfd">
</details>

<details> <summary>After the change: </summary>
 
<img width="972" alt="Screenshot 2023-12-15 at 09 50 52" src="https://github.com/grafana/grafana/assets/33540275/2476dfbd-07fa-4bcd-9568-b1feaa3436bb">
<img width="972" alt="Screenshot 2023-12-15 at 09 50 44" src="https://github.com/grafana/grafana/assets/33540275/ecc36c39-02a9-4f11-b3cf-fc6e644fe1fe">

soniaAguilarPeiron committed:
Hide hysteresis in panel expressions

soniaAguilarPeiron commented:
After a conversation with @yuri-tceretian we agreed on hiding hsyteresis in panel expressions.

JacobsonMT commented:
/deploy-to-hg

yuri-tceretian committed:
Merge branch 'up/main' into yuri-tceretian/expr-hysteresis-alerting

# Conflicts:
#	pkg/tests/api/alerting/testing.go

yuri-tceretian committed:
use new sendRequest function

JacobsonMT commented:
/deploy-to-hg

JacobsonMT reviewed (approved):
Great job! Confirmed working as expected, there are parts I'm not 100% sure on the implementation but considering this is behind a feature flag it's not necessary to dwell on those details at this stage. :rocket:
